### DISCUSSION TITLE
ModuleNotFoundError: No module named 'web3'

### DISCUSSION THREAD

**Discussion by ChinmayaSharma-hue** [smartcontractkit/full-blockchain-solidity-course-py#455] [Q&A] [answered]
Even after I did pip install web3 it is saying that no such module exists when I compile.

**Comment by cromewar** [comment, marked as answer]
You would like to restart vscode to let's it's terminal get a refresh of the new packages, also you must select the python interpreter you have the packages installed on vscode, click on the bottom left corner of vscode and select the correct one.

**Reply by TheCarlins** [reply]
After restarting VSCode and selecting the proper interpreter I still can't resolve this issue. What steps would you recommend to resolve this?

**Reply by TheCarlins** [reply]
After uninstalling and reinstalling pip web3 I come across this error:
WARNING: Ignoring invalid distribution -ip (c:\python310\lib\site-packages)
WARNING: Ignoring invalid distribution - (c:\python310\lib\site-packages)
WARNING: Ignoring invalid distribution -ip (c:\python310\lib\site-packages)
WARNING: Ignoring invalid distribution - (c:\python310\lib\site-packages)  
Collecting web3
  Using cached web3-5.25.0-py3-none-any.whl (485 kB)
Collecting protobuf<4,>=3.10.0
  Using cached protobuf-3.19.1-py2.py3-none-any.whl (162 kB)
Collecting websockets<10,>=9.1
  Using cached websockets-9.1.tar.gz (76 kB)
Collecting ipfshttpclient==0.8.0a2
  Using cached ipfshttpclient-0.8.0a2-py3-none-any.whl (82 kB)
Collecting pywin32>=223
  Using cached pywin32-302-cp310-cp310-win_amd64.whl (9.2 MB)
Collecting hexbytes<1.0.0,>=0.1.0
  Using cached hexbytes-0.2.2-py3-none-any.whl (6.1 kB)
Collecting aiohttp<4,>=3.7.4.post0
  Using cached aiohttp-3.8.1-cp310-cp310-win_amd64.whl (555 kB)
Requirement already satisfied: eth-typing<3.0.0,>=2.0.0 in c:\python310\lib\site-packages (from web3) (2.2.2)
Requirement already satisfied: eth-hash[pycryptodome]<1.0.0,>=0.2.0 in c:\python310\lib\site-packages (from web3) (0.3.2)
Collecting eth-account<0.6.0,>=0.5.6
  Using cached eth_account-0.5.6-py3-none-any.whl (101 kB)
Requirement already satisfied: requests<3.0.0,>=2.16.0 in c:\python310\lib\site-packages (from web3) (2.26.0)
Collecting jsonschema<4.0.0,>=3.2.0
  Using cached jsonschema-3.2.0-py2.py3-none-any.whl (56 kB)
Collecting eth-abi<3.0.0,>=2.0.0b6
  Using cached eth_abi-2.1.1-py3-none-any.whl (27 kB)
Collecting lru-dict<2.0.0,>=1.1.6
  Using cached lru-dict-1.1.7.tar.gz (10 kB)
Requirement already satisfied: eth-utils<2.0.0,>=1.9.5 in c:\python310\lib\site-packages (from web3) (1.10.0)
Collecting multiaddr>=0.0.7
  Using cached multiaddr-0.0.9-py2.py3-none-any.whl (16 kB)
Collecting multidict<7.0,>=4.5
  Using cached multidict-5.2.0-cp310-cp310-win_amd64.whl (45 kB)
Collecting attrs>=17.3.0
  Using cached attrs-21.2.0-py2.py3-none-any.whl (53 kB)
Collecting aiosignal>=1.1.2
  Using cached aiosignal-1.2.0-py3-none-any.whl (8.2 kB)
Collecting frozenlist>=1.1.1
  Using cached frozenlist-1.2.0-cp310-cp310-win_amd64.whl (83 kB)
Collecting async-timeout<5.0,>=4.0.0a3
  Using cached async_timeout-4.0.1-py3-none-any.whl (5.7 kB)
Collecting yarl<2.0,>=1.0
  Using cached yarl-1.7.2-cp310-cp310-win_amd64.whl (122 kB)
Requirement already satisfied: charset-normalizer<3.0,>=2.0 in c:\python310\lib\site-packages (from aiohttp<4,>=3.7.4.post0->web3) (2.0.9)
Requirement already satisfied: typing-extensions>=3.6.5 in c:\python310\lib\site-packages (from async-timeout<5.0,>=4.0.0a3->aiohttp<4,>=3.7.4.post0->web3) (4.0.1)
Requirement already satisfied: parsimonious<0.9.0,>=0.8.0 in c:\python310\lib\site-packages (from eth-abi<3.0.0,>=2.0.0b6->web3) (0.8.1)
Collecting bitarray<1.3.0,>=1.2.1
  Using cached bitarray-1.2.2.tar.gz (48 kB)
Requirement already satisfied: rlp<3,>=1.0.0 in c:\python310\lib\site-packages (from eth-account<0.6.0,>=0.5.6->web3) (2.0.1)
Collecting eth-rlp<2,>=0.1.2
  Using cached eth_rlp-0.2.1-py3-none-any.whl (5.0 kB)
Collecting eth-keys!=0.3.2,<0.4.0,>=0.2.1
  Using cached eth_keys-0.3.3-py3-none-any.whl (20 kB)
Collecting eth-keyfile<0.6.0,>=0.5.0
  Using cached eth_keyfile-0.5.1-py3-none-any.whl (8.3 kB)
Requirement already satisfied: pycryptodome<4,>=3.6.6 in c:\python310\lib\site-packages (from eth-hash[pycryptodome]<1.0.0,>=0.2.0->web3) (3.12.0)
Requirement already satisfied: cytoolz<1.0.0,>=0.9.0 in c:\python310\lib\site-packages (from eth-keyfile<0.6.0,>=0.5.0->eth-account<0.6.0,>=0.5.6->web3) (0.11.2)
Requirement already satisfied: toolz>=0.8.0 in c:\python310\lib\site-packages (from cytoolz<1.0.0,>=0.9.0->eth-keyfile<0.6.0,>=0.5.0->eth-account<0.6.0,>=0.5.6->web3) (0.11.2)
Collecting pyrsistent>=0.14.0
  Using cached pyrsistent-0.18.0-cp310-cp310-win_amd64.whl
Requirement already satisfied: setuptools in c:\python310\lib\site-packages (from jsonschema<4.0.0,>=3.2.0->web3) (57.4.0)
Requirement already satisfied: six>=1.11.0 in c:\python310\lib\site-packages (from jsonschema<4.0.0,>=3.2.0->web3) (1.16.0)
Collecting base58
  Using cached base58-2.1.1-py3-none-any.whl (5.6 kB)
Requirement already satisfied: varint in c:\python310\lib\site-packages (from multiaddr>=0.0.7->ipfshttpclient==0.8.0a2->web3) (1.0.2)
Requirement already satisfied: netaddr in c:\python310\lib\site-packages (from multiaddr>=0.0.7->ipfshttpclient==0.8.0a2->web3) (0.8.0)
Requirement already satisfied: certifi>=2017.4.17 in c:\python310\lib\site-packages (from requests<3.0.0,>=2.16.0->web3) (2021.10.8)
Requirement already satisfied: idna<4,>=2.5 in c:\python310\lib\site-packages (from requests<3.0.0,>=2.16.0->web3) (3.3)
Requirement already satisfied: urllib3<1.27,>=1.21.1 in c:\python310\lib\site-packages (from requests<3.0.0,>=2.16.0->web3) (1.26.7)
Using legacy 'setup.py install' for bitarray, since package 'wheel' is not installed.
Using legacy 'setup.py install' for lru-dict, since package 'wheel' is not installed.
Using legacy 'setup.py install' for websockets, since package 'wheel' is not installed.
WARNING: Ignoring invalid distribution -ip (c:\python310\lib\site-packages)
WARNING: Ignoring invalid distribution - (c:\python310\lib\site-packages)
Installing collected packages: multidict, hexbytes, frozenlist, eth-keys, base58, yarl, pyrsistent, multiaddr, eth-rlp, eth-keyfile, eth-abi, bitarray, attrs, async-timeout, aiosignal, websockets, pywin32, protobuf, lru-dict, jsonschema, ipfshttpclient, eth-account, aiohttp, web3
  WARNING: Failed to write executable - trying to use .deleteme logic
ERROR: Could not install packages due to an OSError: [WinError 2] The system cannot find the file specified: 'C:\\Python310\\Scripts\\base58.exe' -> 'C:\\Python310\\Scripts\\base58.exe.deleteme'

WARNING: Ignoring invalid distribution -ip (c:\python310\lib\site-packages)
WARNING: Ignoring invalid distribution - (c:\python310\lib\site-packages)
WARNING: Ignoring invalid distribution -ip (c:\python310\lib\site-packages)
WARNING: Ignoring invalid distribution - (c:\python310\lib\site-packages)
WARNING: Ignoring invalid distribution -ip (c:\python310\lib\site-packages)
WARNING: Ignoring invalid distribution - (c:\python310\lib\site-packages)
WARNING: You are using pip version 21.2.3; however, version 21.3.1 is available.
You should consider upgrading via the 'C:\Python310\python.exe -m pip install --upgrade pip' command

**Reply by cromewar** [reply]
it seems your python is trying to install all the libraries needed using the ones on cache (as normal) you have some paths to solve this.

1. [delete](https://sourceexample.com/article/en/2a5f7e8460daa23d12e07b8c6adb8275/) python temporary folder, this will force to download the packages again.
2. use other python version, 3.10 is newer but also more unstable I'll suggest you to use 3.8.
3. Use a python virtual env. (install virtualenv using pip)
4. Use WSL (Windows subsystem for linux) and manage all your python stuff there.

Please let us know if this info results useful for you.

Best regards, Cromewar.
